op-node/sources/types.go

```diff
@@ -83,71 +87,175 @@ func (info *HeaderInfo) ReceiptHash() common.Hash {
 	return info.receiptHash
 }
 
-type rpcHeaderCacheInfo struct {
+type rpcHeader struct {
+	ParentHash  common.Hash      `json:"parentHash"`
+	UncleHash   common.Hash      `json:"sha3Uncles"`
+	Coinbase    common.Address   `json:"miner"`
+	Root        common.Hash      `json:"stateRoot"`
+	TxHash      common.Hash      `json:"transactionsRoot"`
+	ReceiptHash common.Hash      `json:"receiptsRoot"`
+	Bloom       eth.Bytes256     `json:"logsBloom"`
+	Difficulty  hexutil.Big      `json:"difficulty"`
+	Number      hexutil.Uint64   `json:"number"`
+	GasLimit    hexutil.Uint64   `json:"gasLimit"`
+	GasUsed     hexutil.Uint64   `json:"gasUsed"`
+	Time        hexutil.Uint64   `json:"timestamp"`
+	Extra       hexutil.Bytes    `json:"extraData"`
+	MixDigest   common.Hash      `json:"mixHash"`
+	Nonce       types.BlockNonce `json:"nonce"`
+
+	// BaseFee was added by EIP-1559 and is ignored in legacy headers.
+	BaseFee *hexutil.Big `json:"baseFeePerGas" rlp:"optional"`
+
+	// untrusted info included by RPC, may have to be checked
 	Hash common.Hash `json:"hash"`
 }
 
-type rpcHeader struct {
-	cache  rpcHeaderCacheInfo
-	header types.Header
+// checkPostMerge checks that the block header meets all criteria to be a valid ExecutionPayloadHeader,
+// see EIP-3675 (block header changes) and EIP-4399 (mixHash usage for prev-randao)
+func (hdr *rpcHeader) checkPostMerge() error {
+	// TODO: the genesis block has a non-zero difficulty number value.
+	// Either this block needs to change, or we special case it. This is not valid w.r.t. EIP-3675.
+	if hdr.Number != 0 && (*big.Int)(&hdr.Difficulty).Cmp(common.Big0) != 0 {
+		return fmt.Errorf("post-merge block header requires zeroed difficulty field, but got: %s", &hdr.Difficulty)
+	}
+	if hdr.Nonce != (types.BlockNonce{}) {
+		return fmt.Errorf("post-merge block header requires zeroed block nonce field, but got: %s", hdr.Nonce)
+	}
+	if hdr.BaseFee == nil {
+		return fmt.Errorf("post-merge block header requires EIP-1559 basefee field, but got %s", hdr.BaseFee)
+	}
+	if len(hdr.Extra) > 32 {
+		return fmt.Errorf("post-merge block header requires 32 or less bytes of extra data, but got %d", len(hdr.Extra))
+	}
+	if hdr.UncleHash != types.EmptyUncleHash {
+		return fmt.Errorf("post-merge block header requires uncle hash to be of empty uncle list, but got %s", hdr.UncleHash)
+	}
+	return nil
```

Comment:
Should we additionally check `MixDigest != 0x000` here?

Reply:
Guessing `0x00` is perfectly valid random hash, but need a way to check if it's actively set I guess.

Reply:
It'll still fit in the execution payload format, and the execution payload attributes are checked, so I don't think we've to be so strict with that attribute as with the others which are not fully compatible with the new `ExecutionPayload` type post-merge.